pkg/event_processor/iworker.go

```diff
@@ -127,12 +132,45 @@ func (ew *eventWorker) parserEvents() []byte {
 func (ew *eventWorker) Run() {
 	for {
 		select {
-		case _ = <-ew.ticker.C:
+		case <-ew.ticker.C:
 			// 输出包
 			if ew.tickerCount > MaxTickerCount {
 				//ew.processor.GetLogger().Printf("eventWorker TickerCount > %d, event closed.", MaxTickerCount)
-				ew.Close()
-				return
+				ew.processor.delWorkerByUUID(ew)
+
+				/*
+					When returned from delWorkerByUUID(), there are two possiblities:
+					1) no routine can touch it.
+					2) one routine can still touch ew because getWorkerByUUID()
+					*happen before* delWorkerByUUID()
+
+					When no routine can touch it (i.e.,ew.IfUsed == false),
+					we just drain the ew.incoming and return.
+
+					When one routine can touch it (i.e.,ew.IfUsed == true), we ensure
+					that we only return after the routine can not touch it
+					(i.e.,ew.IfUsed == false). At this point, we can ensure that no
+					other routine will touch it and send events through the ew.incoming.
+					So, we return.
+
+					Because eworker has been deleted from workqueue after delWorkerByUUID()
+					(ordered by a workqueue lock), at this point, we can ensure that
+					no ew will not be touched even **in the future**. So the return is
+					safe.
+
+				*/
+				for {
+					select {
+					case e := <-ew.incoming:
+						ew.writeEvent(e)
+					default:
+						if ew.IfUsed() {
+							continue
```

Comment:
Does it need to sleep here for a while? Otherwise, will it bring the problem of CPU preemption?

Reply:
I have added the Sleep to address your concern.

But in my opinion, this would not bring the problem of CPU preemption.

Firstly, the total time when ew.used is set to true is expected to be short( just like the spinlock) 

Secondly, according to [Go 1.14 Release Note](https://go.dev/doc/go1.14#runtime):

> Goroutines are now asynchronously preemptible. As a result, loops without function calls no longer potentially deadlock the scheduler or significantly delay garbage collection. This is supported on all platforms except windows/arm, darwin/arm, js/wasm, and plan9/*.

There exists function call in the `for select` loop , and it will not bring problem to deadlock the scheduler or delay GC.

Reply:
From my experience, CPU preemption does occur. However, I have not paid attention to the official documentation of Golang. 

Here is a simple example:
```golang
package main

import (
	"fmt"
)

func main() {
	incoming := make(chan int, 1024)

	for {
		select {
		case e := <-incoming:
			fmt.Println("incoming:", e)
		default:
			if true {
				fmt.Println("default")
				continue
			}
			return
		}
	}
	fmt.Println("end")
}
```

So, I think the `sleep` code here  is necessary.